### DISCUSSION TITLE
waypoint_id missing in OSRM formatted Map Matching API response in tracepoints -> tracepoint -> name

### DISCUSSION THREAD

**Discussion by rishav-raj** [valhalla/valhalla#2782] [Q&A]
Hi,

I am trying to call Map Matching API for a gps trace. I have built valhalla from the source (master branch) and running in my local machine. 

**URL :** http://localhost:8002/trace_route
**Method :** POST
**Request :** 
![Screenshot from 2021-01-20 00-12-52](https://user-images.githubusercontent.com/26870615/105078666-53b64780-5ab4-11eb-9cf8-b16cd86c61be.png)
**Response :**
![Screenshot from 2021-01-20 00-13-54](https://user-images.githubusercontent.com/26870615/105078736-6f215280-5ab4-11eb-9750-3c647f2bed12.png)

**The field "name" in tracepoints -> tracepoint -> name is empty.**

But the same field in OSRM Map Matching response has way_id.
![Screenshot from 2021-01-20 00-16-55](https://user-images.githubusercontent.com/26870615/105079038-d9d28e00-5ab4-11eb-9b67-34db3f64ad54.png)

I need this name field (or way_Id) field for some computation.
Is there any way to get this ? Is there any flag / field that needs to be passed in the request ? Or is this an (intermediate) issue / broken code in master branch ?

I have scoured the web, valhalla documentation, github issues and discuss, but didn't find anything related to this.
Any clue / guidance in this regard will be really helpful !

**Comment by rishav-raj** [comment]
I am new to OpenStreetMaps, and still learning many things about OSRM and Valhalla. This question is invalid, because, while working with OSRM, my team had modified car.lua profile and extracted the .pbf file with the same. That's why I was getting way_id in the "name" field. Apologies for asking wrong question on this forum, but the other question remains unanswered, about how to get way_id in tracepoint object in the response. Please check [this question](https://github.com/valhalla/valhalla/discussions/2785#discussion-1909758) for further details.

**Comment by kevinkreiser** [comment]
@rishav-raj no problem. you cant get back extended info in the `osrm` format. if you want extended info such as the way id, you would need to use the native valhalla api format and specify that you want the way id returned in the response. see the documentation here: https://github.com/valhalla/valhalla/blob/master/docs/api/map-matching/api-reference.md#edge-items

**Reply by rishav-raj** [reply]
Thank you @kevinkreiser for the quick response. Initially, we were evaluating the response of /trace_attributes endpoint. But in order to consume its response, we would have had to change a lot of code, which wasn't feasible at that point of time. So we checked if it was somehow possible to get response from valhalla in OSRM format, and we did find that. The only missing piece of the puzzle is that we want way_id too.

In OSRM, we had actually modified car.lua and used it for extracting .pbf file, to send way_id in the "name" field in tracepoints collection in the match api response. But we couldn't find yet if similar feat is possible in valhalla. If, somehow, we are able to achieve the same in valhalla, our purpose will be served. I urge you to kindly take a look at the below link for my detailed explanation regarding this question.

https://github.com/valhalla/valhalla/discussions/2785

**Comment by kevinkreiser** [comment]
@rishav-raj yes we also use lua to load osm similar to how osrm does it. you should be able to modify `graph.lua` with the same workaround

**Reply by rishav-raj** [reply]
Thank you @kevinkreiser for confirming that we were thinking in the right direction. How do I use `graph.lua` to load osm ?
For example, in the case of OSRM, the below does the trick,
`docker run -t -v "${PWD}:/data" osrm/osrm-backend osrm-extract -p /opt/car.lua /data/berlin-latest.osm.pbf`

But in the case of valhalla, we do this to load the osm data,
`valhalla_build_tiles -c valhalla.json switzerland-latest.osm.pbf liechtenstein-latest.osm.pbf`
And I don't think valhalla_build_tiles accepts lua profile. I checked if there's any option to pass lua file path, but all I could find is the "-c " option for the configuration file.

Please see the below screenshot for osrm-extract :
![Screenshot from 2021-01-22 00-05-30](https://user-images.githubusercontent.com/26870615/105396185-98bdb380-5c45-11eb-9606-31e18e397362.png)

And here's the options with valhalla_build_tiles :
![Screenshot from 2021-01-22 00-06-51](https://user-images.githubusercontent.com/26870615/105396330-c571cb00-5c45-11eb-8921-00eb01acaa91.png)

**Reply by nilsnolde** [reply]
easiest is to recompile the valhalla_build_tiles target with the modified lua

**Reply by nilsnolde** [reply]
nit: pls don't share screenshots, paste code

**Reply by amolskh** [reply]
> @rishav-raj yes we also use lua to load osm similar to how osrm does it. you should be able to modify `graph.lua` with the same workaround

Could you point us to sample modifications that we need to do in graph.lua ?
I see at few places we are setting name https://github.com/valhalla/valhalla/blob/master/lua/graph.lua#L1458
I tried changing
kv["name"] = kv["id"]

but didnt work for us.. Which will be appropriate place for extracting id in lua

**Comment by kevinkreiser** [comment]
you can also make a copy of it, modify it, and then point to it in your valhalla.json config file: https://github.com/valhalla/valhalla/blob/master/src/mjolnir/pbfgraphparser.cc#L790

```
{
"mjolnir":{
  "graph_lua_name": "/path/to/your/edited/graph.lua",
  ...
```
